src/libsyntax_ext/proc_macro_registrar.rs

```diff
@@ -232,6 +190,126 @@ impl<'a> Visitor<'a> for CollectCustomDerives<'a> {
+        if let Some(_) = attr.meta_item_list() {
+            self.handler.span_err(attr.span, "`#[proc_macro_attribute]` attribute
+                cannot contain any meta items");
+            return;
+        }
+
+        if self.in_root && item.vis == ast::Visibility::Public {
+            self.attr_macros.push(AttrProcMacro {
+                span: item.span,
+                function_name: item.ident,
+            });
+        } else {
+            let msg = if !self.in_root {
+                "functions tagged with `#[proc_macro_attribute]` must \
+                 currently reside in the root of the crate"
+            } else {
+                "functions tagged with `#[proc_macro_attribute]` must be `pub`"
+            };
+            self.handler.span_err(item.span, msg);
+        }
+    }
+}
+
+impl<'a> Visitor<'a> for CollectProcMacros<'a> {
+    fn visit_item(&mut self, item: &'a ast::Item) {
+        // First up, make sure we're checking a bare function. If we're not then
+        // we're just not interested in this item.
+        //
+        // If we find one, try to locate a `#[proc_macro_derive]` attribute on
+        // it.
+        let is_fn = match item.node {
+            ast::ItemKind::Fn(..) => true,
+            _ => false,
+        };
+
+        enum FoundAttr<'a> {
+            // Found `#[proc_macro_derive]`
+            Derive(&'a ast::Attribute),
+            // Found `#[proc_macro_attribute]`
+            Attribute(&'a ast::Attribute),
+            None
+        }
+
+        let mut found_attr = FoundAttr::None;
+
+        for a in &item.attrs {
+            if a.check_name("proc_macro_derive") {
+                match found_attr {
+                    FoundAttr::Derive(_) => {
+                        self.handler.span_err(a.span(), "Only one `#[proc_macro_derive]` attribute \
+                            is allowed on any given function");
+                        return;
+                    },
+                    FoundAttr::Attribute(_) => {
+                        self.handler.span_err(a.span(), "Cannot combine `#[proc_macro_attribute]` \
+                        and `#[proc_macro_derive]` on the same function");
+                        return;
+                    },
+                    FoundAttr::None => found_attr = FoundAttr::Derive(a),
+                }
+            } else if a.check_name("proc_macro_attribute") {
+                match found_attr {
+                    FoundAttr::Attribute(_) => {
+                        self.handler.span_err(a.span(), "Only one `#[proc_macro_attribute]` \
+                        attribute is allowed on any given function");
+                        return;
+                    },
+                    FoundAttr::Derive(_) => {
+                        self.handler.span_err(a.span(), "Cannot combine `#[proc_macro_attribute]`
+                        and `#[proc_macro_derive]` on the same function");
```

Comment:
Then we wouldn't have to repeat these error messages for `"proc_macro_derive"` and `"proc_macro_attribute"` -- we would just error `if let Some(found_attr)` and use `a.name() == found_attr.name()` to decide between the errors.